Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -1432,6 +1435,87 @@ void Document::set_visited_link_color(Color color)
+            auto context = CSS::Parser::ParsingContext { document() };
+            auto parsed = parse_css_value(context, content.value(), CSS::PropertyID::ColorScheme);
+
+            // 2. If parsed is a valid CSS 'color-scheme' property value, then return parsed.
+            if (!parsed.is_null() && parsed->is_color_scheme()) {
+                m_supported_color_schemes = parsed->as_color_scheme().schemes();
+                return TraversalDecision::Break;
+            }
+        }
+
+        return TraversalDecision::Continue;
+    });
+
+    // 3. Return null.
+}
+
+// https://html.spec.whatwg.org/multipage/semantics.html#meta-theme-color
+void Document::obtain_theme_color()
+{
+    // 1. Let candidate elements be the list of all meta elements that meet the following criteria, in tree order:
+    for_each_in_subtree_of_type<HTML::HTMLMetaElement>([&](HTML::HTMLMetaElement& element) {
+        //     * the element is in a document tree;
+        //     * the element has a name attribute, whose value is an ASCII case-insensitive match for theme-color; and
+        //     * the element has a content attribute.
+
+        // 2. For each element in candidate elements:
+        auto content = element.attribute(HTML::AttributeNames::content);
+        if (element.name().has_value() && element.name()->equals_ignoring_ascii_case("theme-color"sv) && content.has_value()) {
+            // 1. If element has a media attribute and the value of element's media attribute does not match the environment, then continue.
+            auto context = CSS::Parser::ParsingContext { document() };
+            auto media = element.attribute(HTML::AttributeNames::media);
+            if (media.has_value()) {
+                auto query = parse_media_query(context, media.value());
+                if (window() && !query->evaluate(*window()))
+                    return TraversalDecision::Continue;
+            }
+
+            // 2. Let value be the result of stripping leading and trailing ASCII whitespace from the value of element's content attribute.
+            auto value = content->bytes_as_string_view().trim(Infra::ASCII_WHITESPACE);
+
+            // 3. Let color be the result of parsing value.
+            auto css_value = parse_css_value(context, value, CSS::PropertyID::Color);
+
+            // 4. If color is not failure, then return color.
+            if (!css_value.is_null() && css_value->is_color()) {
+                Optional<Layout::NodeWithStyle const&> root_node;
+                if (html_element())
+                    root_node = *html_element()->layout_node();
+
+                auto color = css_value->to_color(root_node);
+                document().page().client().page_did_change_theme_color(color);
+                return TraversalDecision::Break;
+            }
+        }
+
+        return TraversalDecision::Continue;
+    });
+
+    // 3. Return nothing(the page has no theme color).
```

Comment:
I think this also needs to clear the theme color at the beginning?

Also the theme-color change could be its own commit.

Reply:
Sure. Although I don't know what the default color is and can't test it because it's only used by the AppKit UI.